From 1fec5362bdb04f4f247a39a96dafd0eb291b4c11 Mon Sep 17 00:00:00 2001
From: Michael Sweet <michael.r.sweet@gmail.com>
Date: Tue, 17 Jan 2017 11:59:38 -0500
Subject: Fix rastertopwg line offset bug (Issue #4942)

---
 filter/rastertopwg.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/filter/rastertopwg.c b/filter/rastertopwg.c
index f478ac580..0bb1b3d42 100644
--- a/filter/rastertopwg.c
+++ b/filter/rastertopwg.c
@@ -1,7 +1,7 @@
 /*
  * CUPS raster to PWG raster format filter for CUPS.
  *
- * Copyright 2011, 2014-2016 Apple Inc.
+ * Copyright 2011, 2014-2017 Apple Inc.
  *
  * These coded instructions, statements, and computer programs are the
  * property of Apple Inc. and are protected by Federal copyright law.
@@ -430,6 +430,9 @@ main(int  argc,				/* I - Number of command-line args */
     if (linesize < inheader.cupsBytesPerLine)
       linesize = inheader.cupsBytesPerLine;
 
+    if ((lineoffset + inheader.cupsBytesPerLine) > linesize)
+      lineoffset = linesize - inheader.cupsBytesPerLine;
+
     line = malloc(linesize);
 
     memset(line, white, linesize);
