From 44c5ad66b87ad11cc8d45497034e3ada8f36f09a Mon Sep 17 00:00:00 2001
From: Michael R Sweet <michael.r.sweet@gmail.com>
Date: Wed, 5 Dec 2018 13:02:09 -0500
Subject: Fix -E option (Issue #5440)

---
 systemv/lpadmin.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/systemv/lpadmin.c b/systemv/lpadmin.c
index 6828f8a1d..ee2438c1d 100644
--- a/systemv/lpadmin.c
+++ b/systemv/lpadmin.c
@@ -1387,12 +1387,6 @@ set_printer_options(
   ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI, "printer-uri", NULL, uri);
   ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_NAME, "requesting-user-name", NULL, cupsUser());
 
-  if (enable)
-  {
-    ippAddInteger(request, IPP_TAG_PRINTER, IPP_TAG_ENUM, "printer-state", IPP_PSTATE_IDLE);
-    ippAddBoolean(request, IPP_TAG_PRINTER, "printer-is-accepting-jobs", 1);
-  }
-
  /*
   * Add the options...
   */
@@ -1425,6 +1419,13 @@ set_printer_options(
     ppdfile = NULL;
 
   cupsEncodeOptions2(request, num_options, options, IPP_TAG_OPERATION);
+
+  if (enable)
+  {
+    ippAddInteger(request, IPP_TAG_PRINTER, IPP_TAG_ENUM, "printer-state", IPP_PSTATE_IDLE);
+    ippAddBoolean(request, IPP_TAG_PRINTER, "printer-is-accepting-jobs", 1);
+  }
+
   cupsEncodeOptions2(request, num_options, options, IPP_TAG_PRINTER);
 
   if ((protocol = cupsGetOption("protocol", num_options, options)) != NULL)
