From 7af4943b8c1c7ef9481718978e5b01bb3f958216 Mon Sep 17 00:00:00 2001
From: Michael Sweet <michael.r.sweet@gmail.com>
Date: Sun, 27 Aug 2017 10:58:06 -0400
Subject: Fix sorting bug.

---
 cups/ppd-cache.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cups/ppd-cache.c b/cups/ppd-cache.c
index cc80abe6a..6ca77163c 100644
--- a/cups/ppd-cache.c
+++ b/cups/ppd-cache.c
@@ -4028,7 +4028,7 @@ _ppdCreateFromIPP(char   *buffer,	/* I - Filename buffer */
 
     for (i = 0; i < (count - 1); i ++)
     {
-      for (j = i + 1; j < count; i ++)
+      for (j = i + 1; j < count; j ++)
       {
         int       ix, iy,               /* First X and Y resolution */
                   jx, jy,               /* Second X and Y resolution */
