Description: Color management per queue using colord.
Author: Joe Simon <jsimon383@gmail.com>
Last-Update: 2014-08-29
--- a/cgi-bin/admin.c
+++ b/cgi-bin/admin.c
@@ -1132,6 +1132,11 @@
 				     IPP_TAG_BOOLEAN)) != NULL)
 	  cgiSetVariable("PRINTER_IS_SHARED",
 			 attr->values[0].boolean ? "1" : "0");
+
+        if ((attr = ippFindAttribute(oldinfo, "printer-is-cm-calibrating",
+                                     IPP_TAG_BOOLEAN)) != NULL)
+          cgiSetVariable("PRINTER_IS_CM_CALIBRATING",
+                         attr->values[0].boolean ? "1" : "0");
       }
 
       cgiCopyTemplateLang("modify-printer.tmpl");
@@ -1149,6 +1154,8 @@
 #endif /* __APPLE__ */
         cgiSetVariable("printer_is_shared", "0");
 
+      cgiSetVariable("printer_is_cm_calibrating", "0");
+
       cgiCopyTemplateLang("add-printer.tmpl");
     }
 
@@ -1335,6 +1342,7 @@
     *    printer-info
     *    ppd-name
     *    device-uri
+    *    printer-is-cm-calibrating
     *    printer-is-accepting-jobs
     *    printer-is-shared
     *    printer-state
@@ -1395,6 +1403,10 @@
     ippAddBoolean(request, IPP_TAG_PRINTER, "printer-is-shared",
                   var && (!strcmp(var, "1") || !strcmp(var, "on")));
 
+    var = cgiGetVariable("printer_is_cm_calibrating");
+    ippAddBoolean(request, IPP_TAG_PRINTER, "printer-is-cm-calibrating",
+                  var && (!strcmp(var, "1") || !strcmp(var, "on")));
+
     ippAddInteger(request, IPP_TAG_PRINTER, IPP_TAG_ENUM, "printer-state",
                   IPP_PRINTER_IDLE);
 
--- a/cups/cups.h
+++ b/cups/cups.h
@@ -241,6 +241,7 @@
 					 * @since CUPS 1.4/OS X 10.6@ */
   CUPS_PRINTER_MFP = 0x4000000,		/* Printer with scanning capabilities
 					 * @since CUPS 1.4/OS X 10.6@ */
+  CUPS_PRINTER_CM_OFF = 0x8000000,	/* Printer is not color-managed */
   CUPS_PRINTER_OPTIONS = 0x6fffc	/* ~(CLASS | REMOTE | IMPLICIT |
 					 * DEFAULT | FAX | REJECTING | DELETE |
 					 * NOT_SHARED | AUTHENTICATED |
--- a/cups/dest.c
+++ b/cups/dest.c
@@ -1437,6 +1437,7 @@
 		  "printer-defaults",
 		  "printer-info",
 		  "printer-is-accepting-jobs",
+		  "printer-is-cm-calibrating",
 		  "printer-is-shared",
 		  "printer-location",
 		  "printer-make-and-model",
@@ -1544,7 +1545,8 @@
 	    !strcmp(attr->name, "marker-types") ||
 	    !strcmp(attr->name, "printer-commands") ||
 	    !strcmp(attr->name, "printer-info") ||
-	    !strcmp(attr->name, "printer-is-shared") ||
+	    !strcmp(attr->name, "printer-info") ||
+	    !strcmp(attr->name, "printer-is-cm-calibrating") ||
 	    !strcmp(attr->name, "printer-make-and-model") ||
 	    !strcmp(attr->name, "printer-mandatory-job-attributes") ||
 	    !strcmp(attr->name, "printer-state") ||
@@ -1581,7 +1583,7 @@
               break;
 	    }
 	}
-#endif /* __APPLE__ */
+#endif /* __APPLE__ */        
         else if (!strcmp(attr->name, "printer-name") &&
 	         attr->value_tag == IPP_TAG_NAME)
 	  printer_name = attr->values[0].string.text;
--- a/cups/encode.c
+++ b/cups/encode.c
@@ -274,6 +274,7 @@
   { 0, "printer-error-policy",	IPP_TAG_NAME,		IPP_TAG_PRINTER },
   { 0, "printer-info",		IPP_TAG_TEXT,		IPP_TAG_PRINTER },
   { 0, "printer-is-accepting-jobs", IPP_TAG_BOOLEAN,	IPP_TAG_PRINTER },
+  { 0, "printer-is-cm-calibrating", IPP_TAG_BOOLEAN,	IPP_TAG_PRINTER },
   { 0, "printer-is-shared",	IPP_TAG_BOOLEAN,	IPP_TAG_PRINTER },
   { 0, "printer-location",	IPP_TAG_TEXT,		IPP_TAG_PRINTER },
   { 0, "printer-make-and-model", IPP_TAG_TEXT,		IPP_TAG_PRINTER },
--- a/cups/ipp-support.c
+++ b/cups/ipp-support.c
@@ -1556,6 +1556,7 @@
     "printer-icons",
     "printer-info",
     "printer-is-accepting-jobs",
+    "printer-is-cm-calibrating",	/* CUPS extension */
     "printer-is-shared",		/* CUPS extension */
     "printer-kind",			/* IPP Paid Printing */
     "printer-location",
--- a/scheduler/classes.c
+++ b/scheduler/classes.c
@@ -506,6 +506,27 @@
 	                "Syntax error on line %d of classes.conf.",
 	                linenum);
     }
+    else if (!_cups_strcasecmp(line, "CM-Calibration"))
+    {
+     /*
+      * Set the initial color calibration mode state...
+      */
+
+      if (value &&
+          (!_cups_strcasecmp(value, "yes") ||
+           !_cups_strcasecmp(value, "on") ||
+           !_cups_strcasecmp(value, "true")))
+        p->calibrating = 1;
+      else if (value &&
+               (!_cups_strcasecmp(value, "no") ||
+        	!_cups_strcasecmp(value, "off") ||
+        	!_cups_strcasecmp(value, "false")))
+        p->calibrating = 0;
+      else
+	cupsdLogMessage(CUPSD_LOG_ERROR,
+	                "Syntax error on line %d of classes.conf.",
+	                linenum);
+    }
     else if (!_cups_strcasecmp(line, "Shared"))
     {
      /*
@@ -772,6 +793,11 @@
     else
       cupsFilePuts(fp, "Accepting No\n");
 
+    if (pclass->calibrating)
+      cupsFilePuts(fp, "CM-Calibration Yes\n");
+    else
+      cupsFilePuts(fp, "CM-Calibration No\n");
+
     if (pclass->shared)
       cupsFilePuts(fp, "Shared Yes\n");
     else
--- a/scheduler/cups-lpd.c
+++ b/scheduler/cups-lpd.c
@@ -60,7 +60,8 @@
 			   int num_options, cups_option_t *options);
 static int	get_printer(http_t *http, const char *name, char *dest,
 		            int destsize, cups_option_t **options,
-			    int *accepting, int *shared, ipp_pstate_t *state);
+			    int *accepting, int *calibrating, int *shared, 
+                            ipp_pstate_t *state);
 static int	print_file(http_t *http, int id, const char *filename,
 		           const char *docname, const char *user,
 			   const char *format, int last);
@@ -401,6 +402,7 @@
             int           destsize,	/* I - Size of destination buffer */
 	    cups_option_t **options,	/* O - Printer options */
 	    int           *accepting,	/* O - printer-is-accepting-jobs value */
+            int           *calibrating,	/* O - printer-is-cm-calibrating value */
 	    int           *shared,	/* O - printer-is-shared value */
 	    ipp_pstate_t  *state)	/* O - printer-state value */
 {
@@ -419,6 +421,7 @@
 		{			/* Requested attributes */
 		  "printer-info",
 		  "printer-is-accepting-jobs",
+                  "printer-is-cm-calibrating",
 		  "printer-is-shared",
 		  "printer-name",
 		  "printer-state"
@@ -431,6 +434,8 @@
 
   if (accepting)
     *accepting = 0;
+  if (calibrating)
+    *calibrating = 0;
   if (shared)
     *shared = 0;
   if (state)
@@ -476,11 +481,12 @@
     * using the printer-info values...
     */
 
-    ipp_attribute_t	*accepting_attr,/* printer-is-accepting-jobs */
-			*info_attr,	/* printer-info */
-			*name_attr,	/* printer-name */
-			*shared_attr,	/* printer-is-shared */
-			*state_attr;	/* printer-state */
+    ipp_attribute_t	*accepting_attr,  /* printer-is-accepting-jobs */
+                        *calibrating_attr,/* printer-is-cm-calibrating */
+			*info_attr,	  /* printer-info */
+			*name_attr,	  /* printer-name */
+			*shared_attr,	  /* printer-is-shared */
+			*state_attr;	  /* printer-state */
 
 
     ippDelete(response);
@@ -535,17 +541,21 @@
       * Get all of the attributes for the current printer...
       */
 
-      accepting_attr = NULL;
-      info_attr      = NULL;
-      name_attr      = NULL;
-      shared_attr    = NULL;
-      state_attr     = NULL;
+      accepting_attr  = NULL;
+      calibrating_attr = NULL;
+      info_attr       = NULL;
+      name_attr       = NULL;
+      shared_attr     = NULL;
+      state_attr      = NULL;
 
       while (attr && attr->group_tag == IPP_TAG_PRINTER)
       {
         if (!strcmp(attr->name, "printer-is-accepting-jobs") &&
 	    attr->value_tag == IPP_TAG_BOOLEAN)
 	  accepting_attr = attr;
+        else if (!strcmp(attr->name, "printer-is-cm-calibrating") &&
+	    attr->value_tag == IPP_TAG_BOOLEAN)
+	  calibrating_attr = attr;
 	else if (!strcmp(attr->name, "printer-info") &&
 	         attr->value_tag == IPP_TAG_TEXT)
 	  info_attr = attr;
@@ -574,6 +584,9 @@
 	if (accepting && accepting_attr)
 	  *accepting = accepting_attr->values[0].boolean;
 
+	if (calibrating && calibrating_attr)
+	  *calibrating = calibrating_attr->values[0].boolean;
+
 	if (shared && shared_attr)
 	  *shared = shared_attr->values[0].boolean;
 
@@ -611,6 +624,19 @@
 	*accepting = attr->values[0].boolean;
     }
 
+    if (calibrating)
+    {
+      if ((attr = ippFindAttribute(response, "printer-is-cm-calibrating",
+				   IPP_TAG_BOOLEAN)) == NULL) 
+      {
+	syslog(LOG_ERR, "No printer-is-cm-calibrating attribute found in "
+			"response from server!");
+        *calibrating = 1;
+      }
+      else
+	*calibrating = attr->values[0].boolean;
+    }
+
     if (shared)
     {
       if ((attr = ippFindAttribute(response, "printer-is-shared",
@@ -643,7 +669,7 @@
 
   num_options = 0;
 
-  if (options && shared && accepting)
+  if (options && shared && calibrating && accepting)
   {
     if ((cups_serverroot = getenv("CUPS_SERVERROOT")) == NULL)
       cups_serverroot = CUPS_SERVERROOT;
@@ -786,6 +812,7 @@
 		docname[1024],		/* Document name */
 		dest[256];		/* Printer/class queue */
   int		accepting,		/* printer-is-accepting */
+                calibrating,		/* printer-is-cm-calibrating */
 		shared,			/* printer-is-shared */
 		num_options;		/* Number of options */
   cups_option_t	*options;		/* Options */
@@ -813,7 +840,7 @@
   */
 
   num_options = get_printer(http, queue, dest, sizeof(dest), &options,
-                            &accepting, &shared, NULL);
+                            &accepting, &calibrating, &shared, NULL);
 
   if (num_options < 0 || !accepting || !shared)
   {
@@ -1362,7 +1389,7 @@
   * Get the actual destination name and printer state...
   */
 
-  if (get_printer(http, queue, dest, sizeof(dest), NULL, NULL, NULL, &state))
+  if (get_printer(http, queue, dest, sizeof(dest), NULL, NULL, NULL, NULL, &state))
   {
     syslog(LOG_ERR, "Unable to get printer %s: %s", queue,
            cupsLastErrorString());
--- a/scheduler/ipp.c
+++ b/scheduler/ipp.c
@@ -988,6 +988,16 @@
 		  pclass->accepting ? "Now" : "No longer");
   }
 
+  if ((attr = ippFindAttribute(con->request, "printer-is-cm-calibrating",
+                               IPP_TAG_BOOLEAN)) != NULL)
+  {
+    cupsdLogMessage(CUPSD_LOG_INFO,
+                    "Setting %s printer-is-cm-calibrating to %d (was %d.)",
+                    pclass->name, attr->values[0].boolean, pclass->calibrating);
+
+    pclass->calibrating = attr->values[0].boolean;
+  }
+
   if ((attr = ippFindAttribute(con->request, "printer-is-shared",
                                IPP_TAG_BOOLEAN)) != NULL)
   {
@@ -2491,6 +2501,16 @@
 		  printer->accepting ? "Now" : "No longer");
   }
 
+  if ((attr = ippFindAttribute(con->request, "printer-is-cm-calibrating",
+                                 IPP_TAG_BOOLEAN)) != NULL)
+  {    
+    cupsdLogMessage(CUPSD_LOG_INFO,
+                    "Setting %s printer-is-cm-calibrating to %d (was %d.)",
+                    printer->name, attr->values[0].boolean, printer->calibrating);
+
+    printer->calibrating = attr->values[0].boolean;
+  }
+
   if ((attr = ippFindAttribute(con->request, "printer-is-shared",
                                IPP_TAG_BOOLEAN)) != NULL)
   {
@@ -4900,6 +4920,10 @@
     ippAddBoolean(con->response, IPP_TAG_PRINTER, "printer-is-accepting-jobs",
                   printer->accepting);
 
+  if (!ra || cupsArrayFind(ra, "printer-is-cm-calibrating"))
+    ippAddBoolean(con->response, IPP_TAG_PRINTER, "printer-is-cm-calibrating",
+                  printer->calibrating);
+
   if (!ra || cupsArrayFind(ra, "printer-is-shared"))
     ippAddBoolean(con->response, IPP_TAG_PRINTER, "printer-is-shared",
                   printer->shared);
@@ -4949,6 +4973,9 @@
     if (!printer->accepting)
       type |= CUPS_PRINTER_REJECTING;
 
+    if (!printer->calibrating)
+      type |= CUPS_PRINTER_CM_OFF;
+
     if (!printer->shared)
       type |= CUPS_PRINTER_NOT_SHARED;
 
--- a/scheduler/printers.c
+++ b/scheduler/printers.c
@@ -95,11 +95,12 @@
                                              uuid, sizeof(uuid)));
   cupsdSetDeviceURI(p, "file:///dev/null");
 
-  p->state      = IPP_PRINTER_STOPPED;
-  p->state_time = time(NULL);
-  p->accepting  = 0;
-  p->shared     = DefaultShared;
-  p->filetype   = mimeAddType(MimeDatabase, "printer", name);
+  p->state       = IPP_PRINTER_STOPPED;
+  p->state_time  = time(NULL);
+  p->accepting   = 0;
+  p->calibrating = 0;
+  p->shared      = DefaultShared;
+  p->filetype    = mimeAddType(MimeDatabase, "printer", name);
 
   cupsdSetString(&p->job_sheets[0], "none");
   cupsdSetString(&p->job_sheets[1], "none");
@@ -1081,6 +1082,26 @@
 	cupsdLogMessage(CUPSD_LOG_ERROR,
 	                "Syntax error on line %d of printers.conf.", linenum);
     }
+    else if (!_cups_strcasecmp(line, "CM-Calibration"))
+    {
+     /*
+      * Set the initial color calibration mode state...
+      */
+
+      if (value &&
+          (!_cups_strcasecmp(value, "yes") ||
+           !_cups_strcasecmp(value, "on") ||
+           !_cups_strcasecmp(value, "true")))
+        p->calibrating = 1;
+      else if (value &&
+               (!_cups_strcasecmp(value, "no") ||
+        	!_cups_strcasecmp(value, "off") ||
+        	!_cups_strcasecmp(value, "false")))
+        p->calibrating = 0;
+      else
+	cupsdLogMessage(CUPSD_LOG_ERROR,
+	                "Syntax error on line %d of printers.conf.", linenum);
+    }
     else if (!_cups_strcasecmp(line, "Type"))
     {
       if (value)
@@ -1443,6 +1464,11 @@
     else
       cupsFilePuts(fp, "Accepting No\n");
 
+    if (printer->calibrating)
+      cupsFilePuts(fp, "CM-Calibration Yes\n");
+    else
+      cupsFilePuts(fp, "CM-Calibration No\n");
+
     if (printer->shared)
       cupsFilePuts(fp, "Shared Yes\n");
     else
--- a/scheduler/printers.h
+++ b/scheduler/printers.h
@@ -69,6 +69,7 @@
 		*op_policy,		/* Operation policy name */
 		*error_policy;		/* Error policy */
   cupsd_policy_t *op_policy_ptr;	/* Pointer to operation policy */
+  int		calibrating;		/* Color-calibration Mode? */
   int		shared;			/* Shared? */
   int		accepting;		/* Accepting jobs? */
   int		holding_new_jobs;	/* Holding new jobs for printing? */
--- a/templates/add-printer.tmpl
+++ b/templates/add-printer.tmpl
@@ -33,6 +33,11 @@
 <TD><INPUT TYPE="HIDDEN" NAME="DEVICE_URI" VALUE="{device_uri}">{device_uri}</TD>
 </TR>
 <TR>
+<TH CLASS="label">Color Calibration Mode:</TH>
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_CM_CALIBRATING" {PRINTER_IS_CM_CALIBRATING=1?CHECKED:}>
+Enabled</TD>
+</TR>
+<TR>
 <TH CLASS="label">Sharing:</TH>
 <TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Share This Printer</TD>
@@ -44,4 +49,4 @@
 </TABLE>
 
 </FORM>
-</DIV>
\ No newline at end of file
+</DIV>
--- a/templates/choose-make.tmpl
+++ b/templates/choose-make.tmpl
@@ -25,6 +25,10 @@
 <TD><INPUT TYPE="HIDDEN" NAME="PRINTER_LOCATION" VALUE="{printer_location}">{printer_location}</TD>
 </TR>
 <TR>
+<TH CLASS="label">Color Calibration Mode:</TH>
+<TD><INPUT TYPE="HIDDEN" NAME="PRINTER_IS_CM_CALIBRATING" VALUE="{?printer_is_cm_calibrating}">{?printer_is_cm_calibrating=?Not:{?printer_is_cm_calibrating=0?Not:}} Enabled</TD>
+</TR>
+<TR>
 <TH CLASS="label">Connection:</TH>
 <TD><INPUT TYPE="HIDDEN" NAME="DEVICE_URI" VALUE="{device_uri}">{device_uri}</TD>
 </TR>
@@ -61,4 +65,4 @@
 </TABLE>
 
 </FORM>
-</DIV>
\ No newline at end of file
+</DIV>
--- a/templates/choose-model.tmpl
+++ b/templates/choose-model.tmpl
@@ -24,6 +24,10 @@
 <TD><INPUT TYPE="HIDDEN" NAME="PRINTER_LOCATION" VALUE="{printer_location}">{printer_location}</TD>
 </TR>
 <TR>
+<TH CLASS="label">Color Calibration Mode:</TH>
+<TD><INPUT TYPE="HIDDEN" NAME="PRINTER_IS_CM_CALIBRATING" VALUE="{?printer_is_cm_calibrating}">{?printer_is_cm_calibrating=?Not:{?printer_is_cm_calibrating=0?Not:}} Enabled</TD>
+</TR>
+<TR>
 <TH CLASS="label">Connection:</TH>
 <TD><INPUT TYPE="HIDDEN" NAME="DEVICE_URI" VALUE="{device_uri}">{device_uri}</TD>
 </TR>
@@ -57,4 +61,4 @@
 </TABLE>
 
 </FORM>
-</DIV>
\ No newline at end of file
+</DIV>
--- a/templates/modify-printer.tmpl
+++ b/templates/modify-printer.tmpl
@@ -27,12 +27,16 @@
 <TD><INPUT TYPE="HIDDEN" NAME="DEVICE_URI" VALUE="{device_uri}">{device_uri}</TD>
 </TR>
 <TR>
+<TH CLASS="label">Color Calibration Mode:</TH>
+<TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_CM_CALIBRATING" {PRINTER_IS_CM_CALIBRATING=1?CHECKED:}>
+Enabled</TD>
+</TR>
+<TR>
 <TH CLASS="label">Sharing:</TH>
 <TD><INPUT TYPE="CHECKBOX" NAME="PRINTER_IS_SHARED" {PRINTER_IS_SHARED=1?CHECKED:}>
 Share This Printer</TD>
 </TR>
 <TR>
-<TR>
 <TD></TD>
 <TD><INPUT TYPE="SUBMIT" VALUE="Continue"></TD>
 </TR>
--- a/templates/printer.tmpl
+++ b/templates/printer.tmpl
@@ -3,7 +3,7 @@
 <H2 CLASS="title"><A HREF="{printer_uri_supported}">{printer_name}</A>
 ({printer_state=3?Idle:{printer_state=4?Processing:Paused}},
 {printer_is_accepting_jobs=0?Rejecting Jobs:Accepting Jobs},
-{server_is_sharing_printers=0?Not:{printer_is_shared=0?Not:}} Shared{default_name={printer_name}?, Server Default:})</H2>
+{server_is_sharing_printers=0?Not:{printer_is_shared=0?Not:}} Shared{default_name={printer_name}?, Server Default:}  {printer_is_cm_calibrating=1?, Not Color-Managed:})</H2>
 
 <FORM METHOD="POST" ACTION="{printer_uri_supported}" NAME="maintenance">
 <INPUT TYPE="HIDDEN" NAME="org.cups.sid" VALUE="{$org.cups.sid}">
@@ -44,4 +44,4 @@
 {sides_default?sides={sides_default}:}</TD></TR>
 </TABLE>
 
-</DIV>
\ No newline at end of file
+</DIV>
