Author: till.kamppeter@gmail.com

--- a/filter/pdftops.c
+++ b/filter/pdftops.c
@@ -312,7 +312,9 @@
     }
     else
 #ifdef HAVE_PDFTOPS
-      pdf_argv[pdf_argc++] = (char *)"-level3";
+      /* Do not emit PS Level 3 with Poppler, some HP PostScript printers
+         do not like it. See https://bugs.launchpad.net/bugs/277404. */
+      pdf_argv[pdf_argc++] = (char *)"-level2";
 #else
       pdf_argv[pdf_argc++] = (char *)"-dLanguageLevel=3";
 #endif /* HAVE_PDFTOPS */
