Bug: http://www.cups.org/str.php?L4033
Bug-Ubuntu: https://launchpad.net/bugs/953321

Index: cups-1.5.2/cups/language.c
===================================================================
--- cups-1.5.2.orig/cups/language.c	2011-11-01 07:06:15.000000000 +0100
+++ cups-1.5.2/cups/language.c	2012-03-13 10:10:38.616940739 +0100
@@ -979,7 +979,22 @@
       */
 
       if (m)
-        cupsArrayAdd(a, m);
+      {
+        if (m->str[0])
+        {
+          cupsArrayAdd(a, m);
+        }
+        else
+        {
+         /*
+          * Translation is empty, don't add it... (STR #4033)
+          */
+
+          free(m->id);
+          free(m->str);
+          free(m);
+        }
+      }
 
      /*
       * Create a new message with the given msgid string...
@@ -1057,7 +1072,22 @@
   */
 
   if (m)
-    cupsArrayAdd(a, m);
+  {
+    if (m->str[0])
+    {
+      cupsArrayAdd(a, m);
+    }
+    else
+    {
+     /*
+      * Translation is empty, don't add it... (STR #4033)
+      */
+
+      free(m->id);
+      free(m->str);
+      free(m);
+    }
+  }
 
  /*
   * Close the message catalog file and return the new array...
