Description: Support gzipped charmaps.
Author: Martin Pitt <martin.pitt@ubuntu.com>

Index: cups-1.4.7/cups/transcode.c
===================================================================
--- cups-1.4.7.orig/cups/transcode.c	2011-07-14 13:28:48.099799326 +0200
+++ cups-1.4.7/cups/transcode.c	2011-07-14 13:29:52.079800414 +0200
@@ -1451,6 +1451,7 @@
   cups_ucs2_t	*crow;			/* Pointer to UCS-2 row in 'char2uni' */
   cups_sbcs_t	*srow;			/* Pointer to SBCS row in 'uni2char' */
   char		line[256];		/* Line from charset map file */
+  char		gzfilename[1027];	/* Filename with .gz suffix */
 
 
  /*
@@ -1477,9 +1478,14 @@
 
   if ((fp = cupsFileOpen(filename, "r")) == NULL)
   {
-    DEBUG_printf(("8get_sbcs_charmap: Returning NULL (%s)", strerror(errno)));
+    snprintf (gzfilename, sizeof(gzfilename), "%s.gz", filename);
 
-    return (NULL);
+    if ((fp = cupsFileOpen(gzfilename, "r")) == NULL)
+    {
+      DEBUG_printf(("8get_sbcs_charmap: Returning NULL (%s)", strerror(errno)));
+
+      return (NULL);
+    }
   }
 
  /*
@@ -1603,6 +1609,7 @@
   char		line[256];		/* Line from charset map file */
   int		i;			/* Loop variable */
   int		legacy;			/* 32-bit legacy char */
+  char		gzfilename[1027];	/* Filename with .gz suffix */
 
 
   DEBUG_printf(("7get_vbcs_charmap(encoding=%d, filename=\"%s\")\n",
@@ -1629,9 +1636,14 @@
 
   if ((fp = cupsFileOpen(filename, "r")) == NULL)
   {
-    DEBUG_printf(("8get_vbcs_charmap: Returning NULL (%s)", strerror(errno)));
+    snprintf (gzfilename, sizeof(gzfilename), "%s.gz", filename);
 
-    return (NULL);
+    if ((fp = cupsFileOpen(gzfilename, "r")) == NULL)
+    {
+      DEBUG_printf(("8get_vbcs_charmap: Returning NULL (%s)", strerror(errno)));
+
+      return (NULL);
+    }
   }
 
  /*
