Description: Before this is tested, 4 subscriptions are recorded, not 3.
Origin: vendor
Author: Didier Raboud <odyx@debian.org>
Last-Update: 2012-10-13
--- a/test/run-stp-tests.sh
+++ b/test/run-stp-tests.sh
@@ -430,7 +430,7 @@
 RequestRoot /tmp/cups-$user/spool
 TempDir /tmp/cups-$user/spool/temp
 PidFile /tmp/cups-$user/cupsd.pid
-MaxSubscriptions 3
+MaxSubscriptions 4
 MaxLogSize 0
 AccessLog /tmp/cups-$user/log/access_log
 ErrorLog /tmp/cups-$user/log/error_log
