Description: Add quirk rule to make Xerox WorkCentre 3220 work
Author: Till Kamppeter <till.kamppeter@gmail.com>
Bug-Upstream: https://github.com/apple/cups/issues/4789
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1406203
--- a/backend/org.cups.usb-quirks
+++ b/backend/org.cups.usb-quirks
@@ -121,6 +121,9 @@
 # Xerox WorkCentre 3210 https://bugs.launchpad.net/bugs/1102470
 0x0924 0x4293 no-reattach
 
+# Xerox WorkCentre 3220 https://bugs.launchpad.net/bugs/1406203 https://github.com/apple/cups/issues/4789
+0x0924 0x4294 no-reattach
+
 # QinHeng Electronics CH340S (USB -> Parallel adapter), https://bugs.launchpad.net/bugs/1000253
 0x1a86 0x7584 no-reattach
 
