Description: Add quirk rule to make Lexmark C540n work
Author: Till Kamppeter <till.kamppeter@gmail.com>
Bug-Upstream: https://github.com/apple/cups/issues/4778
--- a/backend/org.cups.usb-quirks
+++ b/backend/org.cups.usb-quirks
@@ -220,6 +220,9 @@
 # Lexmark E238 (STR #4448)
 0x043d 0x009a no-reattach
 
+# Lexmark C540n (STR #4778)
+0x043d 0x0139 no-reattach
+
 # Canon MX310 (STR #4482)
 0x04a9 0x1728 unidir
 
